lib/ain-grpc/src/rpc/eth.rs

```diff
@@ -984,160 +983,139 @@ impl MetachainRPCServer for MetachainRPCModule {
-            return Err(RPCError::InvalidBlockInput.into());
-        }
-
-        let block_numbers = match input.block_hash {
-            None => {
-                // use fromBlock-toBlock
-                let mut block_number = self.get_block(input.from_block)?.header.number;
-                let to_block_number = self.get_block(input.to_block)?.header.number;
-                let mut block_numbers = Vec::new();
-                if block_number > to_block_number {
-                    return Err(RPCError::FromBlockGreaterThanToBlock.into());
-                }
-
-                while block_number <= to_block_number {
-                    block_numbers.push(block_number);
-                    block_number = block_number
-                        .checked_add(U256::one())
-                        .ok_or(RPCError::ValueOverflow)?;
-                }
-                Ok::<Vec<U256>, Error>(block_numbers)
+        let from_block = if input.from_block.is_some() {
+            if let Some(BlockNumber::Num(block_num)) = input.from_block {
+                // Allow future block number to be specified
+                Some(U256::from(block_num))
+            } else {
+                Some(self.get_block(input.from_block)?.header.number)
             }
-            Some(block_hash) => {
-                let block_number = self
-                    .handler
-                    .storage
-                    .get_block_by_hash(&block_hash)
-                    .map_err(to_custom_err)?
-                    .ok_or(RPCError::BlockNotFound)?
-                    .header
-                    .number;
-                Ok(vec![block_number])
+        } else {
+            None
+        };
+        let to_block = if input.to_block.is_some() {
+            if let Some(BlockNumber::Num(block_num)) = input.to_block {
+                // Allow future block number to be specified
+                Some(U256::from(block_num))
+            } else {
+                Some(self.get_block(input.to_block)?.header.number)
             }
-        }?;
-
-        let logs_result = block_numbers
-            .into_iter()
-            .map(|block_number| {
-                self.handler
-                    .logs
-                    .get_logs(&input.address, &input.topics, block_number)
-                    .map_err(to_custom_err)
-            })
-            .collect::<RpcResult<Vec<Vec<_>>>>()?
-            .into_iter()
-            .flatten()
-            .map(LogResult::from)
-            .collect::<Vec<_>>();
-        Ok(logs_result)
+        } else {
+            None
+        };
+        let topics = if let Some(topics) = input.topics {
+            match topics {
+                LogRequestTopics::VecOfHashes(inputs) => {
+                    Some(inputs.into_iter().map(|input| vec![input]).collect())
+                }
+                LogRequestTopics::VecOfHashVecs(inputs) => Some(inputs),
+            }
+        } else {
+            None
```

Comment:
Same as above

Reply:
Refactored topics pipeline to use map in [e7ae914](https://github.com/DeFiCh/ain/commit/e7ae914e0564533816ef18415a3266cd5fe2b413)